File: Codeforces Round #798 (Div. 2)/C.py
from functools import lru_cache
from collections import Counter, defaultdict
import math
from heapq import *

import sys
def get_int(): return int(input())
def get_ints(): return map(int, sys.stdin.readline().strip().split())
def get_list(): return list(map(int, sys.stdin.readline().strip().split()))
def get_string(): return sys.stdin.readline().strip()

def main():
    testcases = get_int()
    for i in range(testcases):
        n = get_int()
        tree = [[] for j in range(n)]
        for j in range(n-1):
            u, v = get_ints()
            tree[u-1].append(v-1)
            tree[v-1].append(u-1)
        print(solve(tree, n))


def solve(tree, n):
    count_child_node = [0 for i in range(n)]
    def count(node, parent):
        result = 0
        tree[node] = list(filter(lambda ele: ele != parent, tree[node]))
        for child in tree[node]:
            if child == parent:
                continue
            result += 1
            result += count(child, node)
        count_child_node[node] = result
        return result
    count(0, -1)

    @lru_cache(None)
    def dfs(node):
        children = tree[node]
        if len(children) == 0:
            return 0
        elif len(children) == 1:
            return count_child_node[children[0]]
        elif len(children) == 2:
            return max(count_child_node[children[0]] + dfs(children[1]),
                    count_child_node[children[1]] + dfs(children[0]))
        else:
            return 0

    return dfs(0)

if __name__ == "__main__":
    main()

File: Codeforces Round #798 (Div. 2)/C.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
)

var reader *bufio.Reader = bufio.NewReader(os.Stdin)
var writer *bufio.Writer = bufio.NewWriter(os.Stdout)

func printf(f string, a ...interface{}) { fmt.Fprintf(writer, f, a...) }
func print(a ...interface{})            { fmt.Println(a...) }
func scanf(f string, a ...interface{})  { fmt.Fscanf(reader, f+"\n", a...) }

func main() {
	defer writer.Flush()

	var testcases int
	scanf("%d", &testcases)
	for i := 0; i < testcases; i++ {
		var n int
		scanf("%d", &n)
		tree := make([][]int, n)
		for j := 0; j < n; j++ {
			tree[j] = make([]int, 0)
		}
		for j := 0; j < n-1; j++ {
			var u, v int
			scanf("%d %d", &u, &v)
			tree[u-1] = append(tree[u-1], v-1)
			tree[v-1] = append(tree[v-1], u-1)
		}
		print(solve(tree, n))
	}
}

func max(nums ...int) int {
	result := math.MinInt32
	for _, num := range nums {
		if num > result {
			result = num
		}
	}
	return result
}

func solve(tree [][]int, n int) int {
	count := make([]int, n)
	var count_child_node func(int, int) int
	count_child_node = func(node, parent int) int {
		children := make([]int, 0)
		result := 0
		for _, child := range tree[node] {
			if child == parent {
				continue
			}
			result += count_child_node(child, node)
			result += 1
			children = append(children, child)
		}
		tree[node] = children
		count[node] = result
		return result
	}
	count_child_node(0, -1)

	var dfs func(int) int
	dfs = func(node int) int {
		children := tree[node]
		switch len(children) {
		case 0:
			return 0
		case 1:
			return count[tree[node][0]]
		case 2:
			return max(count[children[0]]+dfs(children[1]), count[children[1]]+dfs(children[0]))
		default:
			return 0
		}
	}
	return dfs(0)
}
